### PR TITLE
ARTEMIS-3759 Add mirror controller address filter support

### PR DESCRIPTION
**Pull request by iliya-gr**
Allow replication only certain addresses with mirror controller.
The configuration is similar to cluster address configuration.

### CODE REVIEW

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java [hunk 1]
```diff
@@ -337,4 +368,52 @@ public boolean isMirrorController() {
          return true;
       }
    }
+
+   public static class AddressFilter {
+
+      private final SimpleString[] whiteList;
```
[the remark this answers is not in the document]
Comment (iliya-gr) [reply artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java]: Thanks for pointing this. We definitely need a maven plugin for this :)

**Comment by clebertsuconic** [comment]
it looks good to me...


I will not merge it now as the release I'm doing today is named 2.21.1... 


I will defer it to 2.22.0 in a couple weeks...


can you squash these commits and push -f?



(git rebase -I, using a fix on the second commit to the first)

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 2]
```diff
@@ -2442,6 +2442,13 @@
             </xsd:documentation>
          </xsd:annotation>
       </xsd:attribute>
+      <xsd:attribute name="address" type="xsd:string" use="optional">
+         <xsd:annotation>
+            <xsd:documentation>
+               name of the address this mirror connection applies to
```
Comment (gemmellr) [inline artemis-server/src/main/resources/schema/artemis-configuration.xsd:2448]: Its not just a simple address name it is taking, so this feels like the wrong attribute name and description. Something conveying the 'filter' nature would seem appropriate. The format of what it does take isnt mentioned , so besides the test usage its unclear how anyone would know. There should be doc updates to e.g https://github.com/apache/activemq-artemis/blob/2.21.0/docs/user-manual/en/amqp-broker-connections.md#mirror-configuration to cover whatever the changes ends up being.

Other 'broker connections' stuff such as the sender+receiver already take "address-match" or "queue-match", as shown in the context of the config change in ConfigurationTest-full-config.xml. It would also seem reasonable to align the terminology between them if possible (and perhaps even the syntax and/or implementation; they are implemented a _completely_ different way currently, which doesnt seem that obvious).

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java [hunk 3]
```diff
@@ -337,4 +368,52 @@ public boolean isMirrorController() {
          return true;
       }
    }
+
+   public static class AddressFilter {
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java:372]: This seems like it could be private, or just be its own class, rather than being public so that AMQPMirrorControllerSourceTest can use it for testing, without ever testing anything general of AMQPMirrorControllerSource itself.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/AMQPReplicaTest.java [hunk 4]
```diff
@@ -473,6 +473,52 @@ public void testNoAddressWithAnnotations() throws Exception {
       }
    }
 
+   @Test
+   public void testAddressFilterMet() throws Exception {
+      addressTest("test", true, 0);
```
Comment (gemmellr) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/AMQPReplicaTest.java:478]: Using a 0 wait seems potentially susceptible to failure, as the mirroring isnt synchronous. I know its only checking for > 0 messages being mirrored, but I'd still make it more robust personally. Perhaps just changing it to wait for them all to be mirrored as we do expect them to quickly be. Actually, no need for sending 200 messages either given it doesnt actually look at any of them, sending a couple is enough, and it should look.

I would also remove the 1 second wait in the 'no match' case and replace it with better verification by other means, e.g monitor the mirroring SnF queue having not had any messages added when the send is done, or just nothing being present for that target even when subsequently-sent matching stuff has already been received from the other, or both

E.g combine this into one test that sends a non-matching and then matching message in that order and then in the reverse order checks the matching consumer does get a message (receive(timeout)) and the non-matching consumer does not (receiveNoWait) from their respective target queues. Faster and more reliable, verification of what actually happened / actual use. (It could perhaps even monitor the mirror SnF queue during and after, verifying nothing got added to the SnF queue for the non-matching case, and is added for the matching case).

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd (hunk 2 above)
Comment (iliya-gr) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd:2448]: This is still questionable for me. I don't know what is the better name for this attribute. Because any *-match name will suggest same behavior as in receiver/sender and I suspect that nobody will want to create mirror for each address. I chose this name because it has same meaning as in the cluster configuration and I didn't want to expand terminology.

So I am still in doubt. I don't think we should align mirror and receiver/sender filter behavior because they have different use cases.

Comment (gemmellr) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd:2448]: I dont think the use case for this configuration is all that different, in this case its about what things you want to mirror from, vs the others being about what you want to send from or recieve to. An address 'match' for a mirror definition certainly doesnt need to mean 'create a seperate mirror for every address' just because it might mean that for e.g a sender. I also presume it couldnt really mean that given each mirror uses a specific queue named based on its parent broker-connections name.

Using *address-match*, *address-pattern(s)*, *address-filter(s)*, or anything to similar effect, all seem equally reasonable and far more appropriate to me than just "address" does for a configuration that generally seems to be expected _not_ to just have a single address in it.

I'm not saying the config has to be the same, but I am saying it seems more obvious, and I dont as yet see any compelling reason it should be entirely different when its basically for doing entirely equivalent things, configuring applicable addresses of interest.

In any case how it is configured and what it does should be documented.

Comment (iliya-gr) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: I agree with all of this, maybe except reusing `address-match`. I think using the same term that has different semantic/behaviour is bad and will lead users to misconfiguration. 

Anyway, I have opted for `address-filter`, but if there are any objections I can change it to anything else.

Comment (gemmellr) [reply artemis-server/src/main/resources/schema/artemis-configuration.xsd]: I agree address-match shouldn't be used if the configuration within it is actually entirely different, as it is now. The comment had the second aspect that its not so obvious to me that it should be entirely different when both things are about governing matched addresses.

File: docs/user-manual/en/amqp-broker-connections.md [hunk 5]
```diff
@@ -100,6 +100,26 @@ Add a `<mirror>` element within the `<amqp-connection>` element to configure mir
 
 The following optional arguments can be utilized:
 
+* `address-filter`: Each amqp connection mirror configuration only applies to addresses (and its queues) that match the specified `address-filter` attribute, that is, mirroring events will only be sent to the target broker for addresses that match this filter. 
+  An address is matched on the mirror amqp connection when it begins with the string specified in this field. The `address-filters` attribute on a mirror amqp connection also supports comma separated lists and an exclude syntax `!`. 
+  To prevent an address from being matched on this mirror amqp connection, prepend an address string with `!`. If this attribute is not specified, then all addresses and their queues will be mirrored.
```
[the remark this answers is not in the document]
Comment (gemmellr) [reply docs/user-manual/en/amqp-broker-connections.md:105]: I would also possibly suggest moving the `address-filter` option down below the others, as they are relatively tiny by comparison.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/AMQPReplicaTest.java [hunk 6]
```diff
@@ -473,6 +473,73 @@ public void testNoAddressWithAnnotations() throws Exception {
...
+         Assert.assertEquals("will receive", message.getBody(String.class));
+         connection.close();
+      }
+
+      {
+         ConnectionFactory factory = CFUtil.createConnectionFactory("AMQP", "tcp://localhost:" + AMQP_PORT);
+         Connection connection = factory.createConnection();
+         Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         connection.start();
+         MessageConsumer consumer = session.createConsumer(session.createQueue("replicated"));
```
[the remark this answers is not in the document]
Comment (iliya-gr) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/AMQPReplicaTest.java]: Indeed, it should be `nonReplicated`. I have added double send to prevent such mistype.

**Comment by clebertsuconic** [comment]
I will spend some time reviewing this tomorrow after I finish the release on 2.22.
it seems in good shape.

**Comment by iliya-gr** [comment]
> I will spend some time reviewing this tomorrow after I finish the release on 2.22.

Do I need to squash commits before?

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java [hunk 7]
```diff
@@ -162,6 +172,12 @@ public void createQueue(QueueConfiguration queueConfiguration) throws Exception
          }
          return;
       }
+      if (ignoreAddress(queueConfiguration.getAddress())) {
+         if (logger.isTraceEnabled()) {
+            logger.trace("Skipping create " + queueConfiguration + ", queue address" + queueConfiguration.getAddress() + " is excluded");
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java:177]: Given the filter does actually have 'exclusions' and it might not actually be excluded, so much as just not-included, I would reword this to avoid saying it is 'excluded' specifically, but instead something more general like 'doesnt match filter'.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java [hunk 8]
```diff
@@ -208,6 +232,13 @@ public void sendMessage(Message message, RoutingContext context, List<MessageRef
          return;
       }
 
+      if (ignoreAddress(context.getAddress(message))) {
+         if (logger.isTraceEnabled()) {
+            logger.trace("server " + server + " is discarding send to excluded address");
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java:237]: In addition to the "excluded" comment, for this one I'd also note it should say _which address_ is in play otherwise its not very helpful.

**Comment by gemmellr** [comment]
> > I will spend some time reviewing this tomorrow after I finish the release on 2.22.
> 
> Do I need to squash commits before?

It can be reviewed without squashing them but its never going to be merged like this without squashing, so I'd suggest just doing so.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java (hunk 8 above)
Comment (iliya-gr) [reply artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java]: Agree. I have changed both messages.

**Comment by clebertsuconic** [comment]
this looks awesome.. merging it.

**Comment by clebertsuconic** [comment]
thanks @iliya-gr and @gemmellr

**Comment by gemmellr** [comment]
Just to be sure; you didnt have any concerns with the different broker-connections bits having a completely different forms of governing applicable addresses?

**Comment by gemmellr** [comment]
The question was for @clebertsuconic, about the long discussion we already had, where I said it is not obvious to me that two broker-connection elements (which could even be on the same broker-connection at this point) should have completely different ways for doing what is largely the same thing, governing the applicable addresses.

**Comment by clebertsuconic** [comment]
@gemmellr can we have a separate discussion on that? I missed it... and I'm not sure I understand it. are you concerned with users messing up the config and having two connections filtering the same address?

**Comment by gemmellr** [comment]
Not that, although I guess its possible.

Mostly it just seems bizarre for elements of the same thing (a broker connection) to support such entirely distinct ways of doing what amounts to almost exactly the same thing (configuring the applicable addresses). One way supports a list of entries, the other doesnt. One supports wildcarding, the other doesnt. One supports exclusions, the other doesnt.
